capsules/src/lpm013m126.rs

```
@@ -0,0 +1,587 @@
+
+impl CommandHeader {
+    /// Formats header for transfer
+    fn encode(&self) -> [u8; 2] {
+        (self.gate_line | ((self.mode as u16) << 10)).to_be_bytes()
+    }
+}
+
+type WritePosition = WriteFrame;
+
+/// Area of the screen to which data is written
+#[derive(Debug, Copy, Clone)]
+struct WriteFrame {
+    row: u16,
+    column: u16,
+    width: u16,
+    height: u16,
+}
+
+/// Internal state of the driver.
+/// Each state can lead to the next one in order of appearance.
+#[derive(Debug, Copy, Clone)]
+enum State {
+    Uninitialized,
+    InitializingPixelMemory,
+    /// COM polarity and internal latch circuits
+    InitializingRest,
+
+    // Normal operation
+    Idle(WriteFrame),
+    Writing(WriteFrame),
+
+    /// This driver is buggy. Resetting will try to recover it.
+    Bug,
+}
+
+pub struct Lpm013m126<'a, A: Alarm<'a>, P: Pin, S: SpiMasterDevice> {
+    spi: &'a S,
+    extcomin: &'a P,
+    disp: &'a P,
+
+    state: Cell<State>,
+
+    /// This is responsible for sending callbacks
+    /// for actions completed in software.
+    deferred_caller: &'a DynamicDeferredCall,
+    command_complete_callback: OptionalCell<DeferredCallHandle>,
+    /// Holds the handle and the pending call parameter.
+    write_complete_callback: OptionalCell<(DeferredCallHandle, Option<Result<(), ErrorCode>>)>,
+
+    /// The HIL requires updates to arbitrary rectangles.
+    /// The display supports only updating entire rows,
+    /// so edges need to be cached.
+    frame_buffer: OptionalCell<FrameBuffer<'static>>,
+
+    client: OptionalCell<&'static dyn ScreenClient>,
+    /// Buffer for incoming pixel data.
+    /// It's not submitted directly anywhere.
+    buffer: TakeCell<'static, [u8]>,
+
+    /// Needed forinit and to flip the EXTCOMIN pin at regular intervals
+    alarm: &'a A,
+}
+
+impl<'a, A: Alarm<'a>, P: Pin, S: SpiMasterDevice> Lpm013m126<'a, A, P, S>
+where
+    Self: 'static,
+{
+    /// Caution: passing `frame_buffer` that's too small will panic.
```

Could we add a check for this instead?

---

Not quite sure what was meant, so I moved the panic to the component.

---

I mean instead of panincing can we just return an error

---

So that's what I did: the new() returns an error, and then the component panics.